include/graphblas/reference/blas3.hpp

```diff
@@ -61,6 +61,200 @@ namespace grb {
 
 	namespace internal {
 
+		template<
+			Descriptor descr = descriptors::no_operation,
+			class Operator,
+			typename Tin,
+			typename RITin, typename CITin, typename NITin,
+			typename Tout,
+			typename RITout, typename CITout, typename NITout
+		>
+		RC select_generic(
+			Matrix< Tout, reference, RITout, CITout, NITout > &out,
+			const Matrix< Tin, reference, RITin, CITin, NITin > &in,
+			const std::pair< RITin, CITin > &in_coordinates,
+			const Operator &op,
+			const Phase &phase
+		) {
+			typedef typename std::conditional<
+					std::is_void< Tin >::value,
+					bool, Tin
+				>::type ValueType;
+
+			constexpr bool crs_only = descr & descriptors::force_row_major;
+			constexpr bool transpose_input = descr & descriptors::transpose_matrix;
+			static_assert( !(crs_only && transpose_input),
+				"The descriptors::force_row_major and descriptors::transpose_matrix "
+				"flags cannot be used simultaneously" );
+
+			RC rc = SUCCESS;
+			const ValueType identity = ValueType();
+			const auto &in_raw = transpose_input ? internal::getCCS( in ) : internal::getCRS( in );
+			auto &out_crs = internal::getCRS( out );
+			auto &out_ccs = internal::getCCS( out );
+			const size_t
+				m = transpose_input ? ncols( in ) : nrows( in ),
+				n = transpose_input ? nrows( in ) : ncols( in );
+			const size_t
+				m_out = nrows( out ),
+				n_out = ncols( out );
+
+			// Check if the dimensions fit
+			if( m != m_out || n != n_out ) {
+				return MISMATCH;
+			}
+
+			if( m == 0 || n == 0 ) {
+				return SUCCESS;
+			}
+
+			if( phase == RESIZE ) {
+				size_t nzc = 0;
+
+#ifdef _H_GRB_REFERENCE_OMP_BLAS3
+				#pragma omp parallel default(none) \
+					shared(out, in_raw, rc) firstprivate(m, op, identity, in_coordinates) \
+					reduction(+:nzc)
```

Comment:
I don't think the OpenMP standard specifies the performance semantics of reduction -- it is likely to be implementation-dependent. Therefore an OpenMP run-time implementation may dynamically allocate memory to facilitate the reduction. Instead relying on the internal ALP/GraphBLAS buffer and explicitly doing the reduction prevents this possibility and ensures the implementation adheres to the performance semantics (which, although presently not specified, is likely to include the promise that no system calls such as malloc or free are made).